dart/lib/src/sentry_envelope_item.dart

```diff
@@ -128,11 +129,21 @@ class SentryEnvelopeItem {
     // an exception.
     try {
       final itemHeader = utf8JsonEncoder.convert(await header.toJson());
-
       final newLine = utf8.encode('\n');
       final data = await dataFactory();
-      // TODO the data copy could be avoided - this would be most significant with attachments.
-      return [...itemHeader, ...newLine, ...data];
+
+      final totalLength = itemHeader.length + newLine.length + data.length;
+
+      final result = Uint8List(totalLength);
+      result.setRange(0, itemHeader.length, itemHeader);
+      result.setRange(
+        itemHeader.length,
+        itemHeader.length + newLine.length,
+        newLine,
+      );
+      result.setRange(itemHeader.length + newLine.length, totalLength, data);
```

Comment:
This is still a copy though. 

How about we restructure the whole thing and get rid of envelopeItemStream() altogether? IIRC it wasn't necessary last time I looked at it.

Reply:
I think you could even get rid of the `SentryEnvelopeItemHeader` class, which would get rid of the awkward relationship between `SentryEnvelopeItemHeader` and `SentryEnvelopeItem`, and probably simplifies the code quite a bit. 

I think the C# SDK also doesn't have two seperate classes for it.

Reply:
I checked and we did use yield here before. This was changed in this [PR](https://github.com/getsentry/sentry-dart/pull/505/files#) @ueman Do you remember if the issue described in this comment did occur?

 >  // Each item needs to be encoded as one unit.
 >  // Otherwise the header already got yielded if the content throws
 >  // an exception.

Because the header, in order to get the data length, it would access have to access the dataFactory, so any error would be thrown on the first yield anyway as far as i understood.

```dart
/// Stream binary data of `Envelope` item.
Stream<List<int>> envelopeItemStream() async* {
  yield utf8.encode(jsonEncode(await header.toJson())); // accessing _CachedItem.getDataLength, therefore pre-caching getData
  yield utf8.encode('\n');
  yield await dataFactory(); // dataFactory == _CachedItem.getData
}
   
class _CachedItem {
  _CachedItem(this._dataFactory);

  final Future<List<int>> Function() _dataFactory;
  List<int>? _data;

  Future<List<int>> getData() async {
    _data ??= await _dataFactory();
    return _data!;
  }

  Future<int> getDataLength() async {
    try {
      return (await getData()).length;
    } catch (_) {
      return -1;
    }
  }
}
```

I checked `sentry-dotnet`. The header it just an dictionary in [`EnvelopeItem.cs`](https://github.com/getsentry/sentry-dotnet/blob/cbe1af4c722b22eb6ee0cbe5c70b25a8d7056d41/src/Sentry/Protocol/Envelopes/EnvelopeItem.cs#L33), where on [iOS](https://github.com/getsentry/sentry-cocoa/blob/1928267073dc3a0d3ce4e9f7912d52404ca5ee77/Sources/Sentry/Public/SentryEnvelopeItemHeader.h#L4) and [Android](https://github.com/getsentry/sentry-java/blob/3d24791bd1b80d3dbf8a849d9f23bf14445902c9/sentry/src/main/java/io/sentry/SentryEnvelopeItemHeader.java#L14) we have classes that mirror the JSON structure.

@ueman Could you elaborate on what you meant by `akward` and how the C# approach would be beneficial here? Just want to understand what you meant so we can decide if this refactor is worth it.

Reply:
I removed the `envelopeItemStream` returning a concatenated list and moved the actual yielding of data into the envelope. Also removed the complexity with `_CachedItem` and the length closure in the `SentryEnvelopeItemHeader`, we don't need this.

Reply:
> Do you remember if the issue described in this comment did occur?
> 
> // Each item needs to be encoded as one unit.
> // Otherwise the header already got yielded if the content throws
> // an exception.

I did run into the issue where I could create the header, but not the item itself, which then resulted in a broken envelop. I don't know of a specific example though.

> Could you elaborate on what you meant by akward and how the C# approach would be beneficial here? Just want to understand what you meant so we can decide if this refactor is worth it.

What I mean by that is that you can probably get rid of the `_CachedItem` and friends, since you no longer need to read size etc twice, since you could do all of that in just one class. I can't really tell whether it's worth it though.